### DISCUSSION TITLE
Optional faer

### DISCUSSION THREAD

**Discussion by albersonmiranda** [extendr/extendr#804] [Q&A]
In [this issue](https://github.com/albersonmiranda/fio/issues/39), @JosiahParry suggested using `faer` feature to facilitate transitioning from and to R and Rust.

But if I do something like:
```
fn foo(
  some_mat: RMatrix<f64>
) -> Mat<f64> {
  
  // convert to faer matrix
  let some_mat_faer = Mat::<f64>::from(some_mat);
  ```

I still get
```
mismatched types
expected struct `faer::Mat<f64>`
   found struct `extendr_api::RArray<f64, [usize; 2]>
```

My dependencies on Cargo.toml:
```
[dependencies]
faer = '*'
rayon = '*'
extendr-api = { git = "https://github.com/extendr/extendr", features = ["faer"] }
```

**Comment by CGMossa** [comment]
I would need to test this more hands-on. I see that we have `impl From<RMatrix<f64>> for Mat<f64>` which should cover it. 

Update: I've just tried to create a package from scratch, and add such a function, and it works.

**Reply by JosiahParry** [reply]
@CGMossa mind sharing it? because I'm actually able to reproduce the issue. I get closer by adding 

```toml
[dependencies]
extendr-api = { git = "https://github.com/extendr/extendr", features = [
    "faer",
] }
faer = '0.18.0'
```

which uses the same version that is in the extendrtests repo. 

Then this actually causes a compilation error 

```
error[E0599]: no method named `row_iter` found for struct `Mat` in the current scope
  --> src/linkages.rs:32:77
   |
32 | ...atrix_faer.row_iter()
   |               ^^^^^^^^ method not found in `Mat<f64>`

error[E0599]: no method named `row_iter` found for struct `Mat` in the current scope
   --> src/linkages.rs:130:6
    |
129 |     let lim_minus_ca: Vec<f64> = leontief_inverse_matrix_faer
    |  ________________________________-
130 | |     .row_iter()
    | |     -^^^^^^^^ method not found in `Mat<f64>`
    | |_____|
    |
```

eeek

**Reply by albersonmiranda** [reply]
@CGMossa ping. Since I can't use Vec<Vec<>>  as input, I don't see a way to work with matrix input without declaring it's dimensions in args.

**Reply by CGMossa** [reply]
I don't really know how to answer this.
In R, the matrix is stored as a flat, 1D, column-wise slice. You can think of it as `&[f64]` or `&[Rfloat]` if you need `NA`-awareness. You can also use `RMatrix`. Basically, you can get access to this flat-representation using `.data` or `.data_mut()` and with that you should be able to do everything.
`Vec<Vec<_>>` is not possible to attain, because that is not how the data is stacked.

**Reply by CGMossa** [reply]
> @CGMossa mind sharing it? because I'm actually able to reproduce the issue. I get closer by adding
> 
I've put it here <https://github.com/CGMossa/faerr>. I'm not sure what is going on, but it is clearly something rust-specific that we have overseen. Current suspects are

* Versions specified in `Cargo.toml`
* Inclusion of a `faer` in the `extendr_api::prelude`
* Lack of running `cargo update`

**Reply by CGMossa** [reply]
There is clearly something going on here:
```rs

faerr/src/rust on  main [!?] is 📦 v0.1.0 via 🦀 v1.78.0 
❯ cargo tree -i faer
  Downloaded bytemuck v1.16.1
  Downloaded 1 crate (48.3 KB) in 0.15s
error: There are multiple `faer` packages in your project, and the specification `faer` is ambiguous.
Please re-run this command with one of the following specifications:
  faer@0.18.2
  faer@0.19.0

faerr/src/rust on  main [!?] is 📦 v0.1.0 via 🦀 v1.78.0 
❯ cargo tree -i faer@0.19.0
faer v0.19.0
└── faerr v0.1.0 (/Users/elea/Documents/GitHub/faerr/src/rust)

faerr/src/rust on  main [!?] is 📦 v0.1.0 via 🦀 v1.78.0 
❯ cargo tree -i faer@0.18.2
faer v0.18.2
└── extendr-api v0.6.0 (https://github.com/extendr/extendr#d2c40148)
    └── faerr v0.1.0 (/Users/elea/Documents/GitHub/faerr/src/rust)
```

The `cargo tree -i <pkg>` sort of yields reverse-dependencies based on current affair.

**Reply by albersonmiranda** [reply]
> I don't really know how to answer this. In R, the matrix is stored as a flat, 1D, column-wise slice. You can think of it as `&[f64]` or `&[Rfloat]` if you need `NA`-awareness. You can also use `RMatrix`. Basically, you can get access to this flat-representation using `.data` or `.data_mut()` and with that you should be able to do everything. `Vec<Vec<_>>` is not possible to attain, because that is not how the data is stacked.

Nevermind, I was trippin. I was trying to use `Mat::from(m)` which is the issue here, but as a workaround it's possible to just
```
mat: RMatrix<f64>
...
let n = mat.nrows();
let m = mat.ncols();

let mat_f = Mat::from_fn(n, m, |row, col| mat[(row, col).into()])
```
Before I was doing exactly what you've said, but only for square matrices:
```
mat: &[f64}

let dim = (mat.len() as f64).sqrt() as usize;
let mat_f = Mat::from_fn(n_fd, m_fd, |row, col| mat[(row, col).into()])
```
